Application Endpoint Discovery API - Update specification according to issues created

jgarciatovar opened the pull request:
#### What type of PR is this?

This PR provides changes in AED API spec including some corrections, minor typo errors, and enhancements proposed.

#### Which issue(s) this PR fixes:

* #366 - updated Device object schema according to Commonalities definition
* #367 - updated Errors and x-correlator parameter/header to Commonalities definition
* #368 - update spec to use EdgeCloudZone object as defined by EAM API
* #362 - Updated 401 error removing AUTHENTICATION required (https://github.com/camaraproject/Commonalities/pull/429 as part of Commonalities 0.6 updates for Fall25)
* Minor typo error on ResourceApplicationEndpoint object
* Added some examples to the API spec
* Added reference to commonalities 0.5 spec (to be updated to 0.6 in later PR)
* Updated operationId to getClosestAppEndpoint which defines better the intent of the operation

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -130,7 +129,7 @@ paths:
         - openId:
             - application-endpoint-discovery:app-endpoints:read
-      operationId: getAppEndpoints
+      operationId: getClosestAppEndpoint
```
why is the operationID singular (endpoint) but the path, response object and openid scope are all plural (endpoints)?

jgarciatovar replied:
Thanks for your comment Kevin!

My view is that Application Endpoint Discovery API should be aligned with Simple Egde Discovery API, returning just one single application endopoint in the response. The assumption is that there will be only one application Instance running per zone for a given AppId, so this API returns the closest (or optimal) endpoint (only one). That's how SED is now defined, returning the closest Zone.

On the other hand, Optimal Edge Discovery API can return an array of zones. That might make sense as there might be multiple zones that comply with requested applicationProfile.

Thus, my opinion is that:
- OED returns array of zones
- AED and SED return single endpoint/zone

What do you think @maheshc01 ?

If you agree, I'll create a separated issue to update AED so it returns single element instead of array.

gunjald replied:
I think we should also define the term "endpoint" just to avoid any confusion and also the cardinality w.r.t an application whether one or more endpoint can be there.

gunjald replied:
A minor suggestion to add FQDN unless there is some reason to not use it: ""Application endpoint : an application server IP address or FQDN published by the Edge Cloud Zone provider."

Kevsy replied:
Yes, good catch @gunjald

Kevsy replied:
Thanks @jgarciatovar - very minor corrections suggested otherwise LGTM

Kevsy replied:
LGTM @jgarciatovar 👍

jgarciatovar commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -181,7 +190,7 @@ paths:
-        - Endpoint Discovery
+        - Application Endpoint Discovery
       summary: |
         Returns the Endpoint of the closest App Instance to user device
```
This concept has been already discussed with @Kevsy on #357 . Note that Simple Edge Discovery API uses the term closest too.

Currently the Application Endpoint Discovery returns the closest endpoint (following the same terminology as Simple Edge Discovery). So we'll keep the term closest (instead of optimal) in the API Definition and other documentation.

In the future, if some enhancements are added to consider other ways to obtain the returned endpoint, we might update that to 'optimal' term.

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -9,29 +9,30 @@ info:
-    Edge Cloud Zone but also providing direct links to your specified
-    application's endpoints within those Edge Cloud Zones.
-    This API intent enables immediate connection to the most relevant
+    Edge Cloud Zone but also providing direct link to your specified
```
```suggestion
    Edge Cloud Zone but also providing the direct link to your specified
```

jgarciatovar replied:
Marked as solved as it is now committed.

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -9,29 +9,30 @@ info:
-    application's endpoints within those Edge Cloud Zones.
-    This API intent enables immediate connection to the most relevant
+    Edge Cloud Zone but also providing direct link to your specified
+    application's endpoint within such Edge Cloud Zone.
```
```suggestion
    application's endpoint within that Edge Cloud Zone.
```

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -295,29 +429,45 @@ components:
       example: "123e4567-e89b-12d3-a456-426614174000"
-    ResourcesapplicationEndpoint:
+
+    ResourcesApplicationEndpoint:
```
Why is it `ResourcesApplicationEndpoint` and not `ApplicationEndpoint`?

jgarciatovar replied:
Marked as solved - updated name in teh API to `ApplicationEndpoint`

jgarciatovar commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -295,29 +429,45 @@ components:
+      description: |
+        An application server IP address:port or FQDN published by
+        the Edge Cloud Zone provider.
       properties:
```
@Kevsy / @gunjald / @maheshc01  - ApplicationEndpoint is now aligned with EWBI GSMA interface. In addition to those parameters, I've added the `EdgeCloudZone` which is 1:1 relation with the Endpoint (and it is neded inside this object to allow multiple endpoints returned in the same response).

Kevsy replied:
Great, thanks @jgarciatovar

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+    networking demands (latency, jitter, bandwidth rates).
+    - `Application Provider:` The provider of the application that is deployed
+    on the Edge Cloud.
+    - `Application Endpoint:` An application server IP address or FQDN, and
```
Is port mandatory? I would have thought it would typically default to 443

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+    port exposed by an application instance to allow connections from clients.
+    - `Application Instance:` An instantiation of an Application running on
+    an Edge Cloud Zone. It is assumed that for each application, a maximum of
+    one instance can run per Edge Cloud Zone.
```
Should we say "logical instance" here to account for load balancing?

jgarciatovar replied:
Agreed. I'll update it in the next commit

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+      fetched via external IP discovery services.
+      - `phoneNumber:` The device's phone number in international E.164 format.
+      - `networkAccessIdentifier:` A unique identifier provided by the network
     operator.
```
```suggestion
    operator (where supported).
```

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+
+    - `applicationEndpoints:` array containing at least one Application Endpoint
+    object including the following parameters:
+      - `port:` TCP or UDP port number exposed by the application instance.
```
As above, should this be mandatory? Or would we assume a default of 443?

jgarciatovar replied:
To be aligned with current definition on EWBI interface, I would leave as mandatory for now.
The nature of the port will depend on the application, and there you can consider any TCP/UDP protocol and port (not just http or https).

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+      - One of the following parameters which define the endpoint:
+        - `ipv4addresses:` List of IPv4 addresses exposed by the application
+        instance.
+        - `ipv6addresses:` List of IPv6 addresses exposed by the application
```
```suggestion
        - `ipv6Addresses:` Array of IPv6 addresses exposed by the application
```

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -463,23 +558,58 @@ components:
+        A single IPv4 address specified in dotted-quad form like 1.2.3.4.
+      example: "198.51.100.1"
+
+    Ipv6Addr:
```
```suggestion
    Ipv6Address:
```

Kevsy reviewed (changes requested):
Some suggestions made inline: including for the use of `Addr` instead of `Address` which will require additional changes not in this PR. So I can raise another PR to make them all consistent once this one is approved.

jgarciatovar commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+    using the Edge Application Management API, and available endpoints have been
+    registered by the Edge Application Registration API.
+
+    If the API is accessed directly from a device within the operator's network,
+    initial requests may be made without explicit device identifiers (see
+    `Identifying the device from the access token` section). However, if this
+    results in a `404 Not Found` error, subsequent requests should include one of
+    the identifiers to ensure accurate localization of the nearest endpoint.
```
Ok. To be removed on the next commit.

jgarciatovar commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
+    If successful, a JSON object is returned on 200 response containing the
+    following data:
+
+    - `applicationEndpoints:` array containing at least one Application Endpoint
+    object including the following parameters:
```
To be added on next commit.

jgarciatovar commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -4,84 +4,146 @@ info:
-    * `Network-Access-Identifier`: A unique identifier provided by the network
+    endpoint(s), enhancing the responsiveness of the service.
+    * If no suitable endpoints are available, the developer can use other APIs
+    to instaintiate new instance and/or register new application endpoints.
```
Ok. To be updated on next commit.

jgarciatovar commented:
> Some suggestions made inline: including for the use of `Addr` instead of `Address` which will require additional changes not in this PR. So I can raise another PR to make them all consistent once this one is approved.

Thanks for all the suggestions @Kevsy  - I've just resolved them and added a commit with the changes. 
I don't understand what you mean by the use of `Addr` instead of `Address`. I've updated `Ipv4Addr` to `Ipv4Address` as suggested in the last commit. Am I missing anything? I don't see which additional changes are needed there. 

@maheshc01 / @gunjald / @Kevsy / @JoseMConde : I think that this PR is now ready to merge, unless someone raises their hand.  @Kevsy performed a very detailed review, and we might need some additional changes before Fall25 meta-release (ICM and Consent Management) that should go in a different PR. 
My suggestion is that I leave this open today for further review, and if no other requests, we'll proceed with merge tomorrow.

Kevsy commented on code/API_definitions/Application_Endpoint_Discovery_API.yaml:
```diff
@@ -206,7 +308,7 @@ components:
       schema:
         type: string
-        format: uuid
+        pattern: ^[a-zA-Z0-9-]{0,55}$
```
```suggestion
        pattern: ^[a-zA-Z0-9-_:;.\/<>{}]{0,256}$
```

Kevsy reviewed (dismissed):
LGTM,  the new x-correlator pattern needs to be applied but let's do that in a separate PR :)

jgarciatovar commented:
@Kevsy - I am running locally spectral and facing some linting errors. So I need to add a new commit before merging this PR. Apologies for inconvenience.

jgarciatovar commented:
@Kevsy - can you please review again and approve? Basically, I had to remove the example because I got an error due to something related with anyOf and/or array of endpoints. Something not mandatory for now that I'll try to add later.
